src/gnatss/ops/posfilter.py

```diff
@@ -75,3 +79,86 @@ def rotation(
     df.drop(columns=[*d_enu_columns, *td_enu_columns], inplace=True)
 
     return df
+
+
+def kalman_filtering(
+    inspvaa_df: DataFrame,
+    insstdeva_df: DataFrame,
+    gps_df: DataFrame,
+) -> DataFrame:
+    """
+    Performs Kalman filtering of the GPS_GEOCENTRIC and GPS_COV_DIAG fields
+
+    Parameters
+    ----------
+    inspvaa_df :  DataFrame
+        Pandas Dataframe containing Antenna enu directions Novatel L1 data
+    insstdeva_df :  DataFrame
+        Pandas Dataframe containing Antenna enu directions std deviation Novatel L1 data
+    gps_df :  DataFrame
+        Pandas Dataframe containing GPS solutions Novatel L1 data
+
+    Returns
+    -------
+    DataFrame
+        Pandas Dataframe containing Time and Kalman filtered GPS_GEOCENTRIC and GPS_COV_DIAG columns
+    """
+    inspvaa_df = inspvaa_df.rename(
+        columns={
+            constants.TIME_J2000: constants.GPS_TIME,  # TODO For merging convenience (So extra cols dont pop up during merge)
+        },
+        errors="raise",
+    )
+    inspvaa_df = inspvaa_df[
+        [
+            constants.GPS_TIME,
+            *constants.ANTENNA_DIRECTIONS,
+        ]
+    ]
+    insstdeva_df = insstdeva_df.rename(
+        columns={
+            constants.TIME_J2000: constants.GPS_TIME,
+        },
+        errors="raise",
+    )
+    insstdeva_df = insstdeva_df[
+        [
+            constants.GPS_TIME,
+            f"{constants.ANTENNA_EASTWARD} std",
+            f"{constants.ANTENNA_NORTHWARD} std",
+            f"{constants.ANTENNA_UPWARD} std",
+        ]
+    ]
+
+    insstdeva_df["v_sden"] = 0.0  # TODO Should I create a constant for these columns?
+    insstdeva_df["v_sdeu"] = 0.0
+    insstdeva_df["v_sdnu"] = 0.0
+
+    gps_df["sdxy"] = np.sqrt(
+        gps_df[constants.GPS_COV_XX] * gps_df[constants.GPS_COV_YY]
+    )
+    gps_df["sdxz"] = np.sqrt(
+        gps_df[constants.GPS_COV_XX] * gps_df[constants.GPS_COV_ZZ]
+    )
+    gps_df["sdyz"] = np.sqrt(
+        gps_df[constants.GPS_COV_YY] * gps_df[constants.GPS_COV_ZZ]
+    )
+
+    merged_df = inspvaa_df.merge(gps_df, on=constants.GPS_TIME, how="left")
+    merged_df = merged_df.merge(insstdeva_df, on=constants.GPS_TIME, how="left")
+    merged_df = merged_df.sort_values(constants.GPS_TIME).reset_index(drop=True)
+
+    first_pos = merged_df[~merged_df.x.isnull()].iloc[0].name
+    merged_df = merged_df.loc[first_pos:].reset_index(drop=True)
+
+    merged_np_array = merged_df.to_numpy()
+    x, P, K, Pp = run_filter_simulation(merged_np_array)
+
+    smoothed_results = DataFrame(
+        x.reshape(x.shape[0], -1),
+        columns=[*constants.GPS_GEOCENTRIC, *constants.GPS_COV_DIAG],
+    )
+    smoothed_results[constants.GPS_TIME] = merged_df[constants.GPS_TIME]
+    # TODO How are the GPS_COV values calculated???
+
+    return smoothed_results

```

Comment:
Just leave to return the 6 cols, but check out the `x`,`y`,`z` against input values